core/src/main/scala/ai/h2o/sparkling/backend/Converter.scala

```diff
@@ -0,0 +1,216 @@
+                                    sparse: Array[Boolean],
+                                    partitions: Seq[Int],
+                                    partitionSizes: Map[Int, Int])(context: TaskContext, it: Iterator[Row]): (Int, Long) = {
+    val chunkIdx = partitions.indexOf(context.partitionId())
+    val partitionSize = partitionSizes(context.partitionId())
+    withResource(
+      new Writer(conf,
+        uploadPlan(chunkIdx),
+        keyName,
+        partitionSize,
+        expectedTypes,
+        chunkIdx,
+        maxVecSizes,
+        sparse)) { writer =>
+      it.foldLeft(0) {
+        case (localRowIdx, row) => sparkRowToH2ORow(row, localRowIdx, writer, elemStartIndices, maxVecSizes)
+      }
+    }
+    (chunkIdx, partitionSize)
+  }
+
+  /**
+   * Converts a single Spark Row to H2O Row with expanded vectors and arrays
+   */
+  private def sparkRowToH2ORow(row: Row, rowIdx: Int, con: Writer, elemStartIndices: Array[Int], maxVecSizes: Array[Int]): Int = {
+    con.startRow(rowIdx)
+    row.schema.fields.zipWithIndex.foreach { case (entry, idxField) =>
+      val idxH2O = elemStartIndices(idxField)
+      if (row.isNullAt(idxField)) {
+        con.putNA(idxH2O, idxField)
+      } else {
+        entry.dataType match {
+          case BooleanType => con.put(idxH2O, row.getBoolean(idxField))
+          case ByteType => con.put(idxH2O, row.getByte(idxField))
+          case ShortType => con.put(idxH2O, row.getShort(idxField))
+          case IntegerType => con.put(idxH2O, row.getInt(idxField))
+          case LongType => con.put(idxH2O, row.getLong(idxField))
+          case FloatType => con.put(idxH2O, row.getFloat(idxField))
+          case _: DecimalType => con.put(idxH2O, row.getDecimal(idxField).doubleValue())
+          case DoubleType => con.put(idxH2O, row.getDouble(idxField))
+          case StringType => con.put(idxH2O, row.getString(idxField))
+          case TimestampType => con.put(idxH2O, row.getAs[java.sql.Timestamp](idxField))
+          case DateType => con.put(idxH2O, row.getAs[java.sql.Date](idxField))
+          case v if ExposeUtils.isMLVectorUDT(v) => con.putVector(idxH2O, row.getAs[ml.linalg.Vector](idxField), maxVecSizes(idxField))
+          case _: mllib.linalg.VectorUDT => con.putVector(idxH2O, row.getAs[mllib.linalg.Vector](idxField), maxVecSizes(idxField))
+          case udt if ExposeUtils.isUDT(udt) => throw new UnsupportedOperationException(s"User defined type is not supported: ${udt.getClass}")
+          case unsupported => throw new UnsupportedOperationException(s"Data of type ${unsupported.getClass} are not supported for the conversion" +
+            s"to H2OFrame.")
+        }
+      }
+    }
+    rowIdx + 1
+  }
+
+  private def scheduleUpload[T](numPartitions: Int, rdd: H2OAwareRDD[T]): UploadPlan = {
+    val hc = H2OContext.ensure("H2OContext needs to be running")
+    val nodes = hc.getH2ONodes()
+    if (hc.getConf.runsInInternalClusterMode) {
```

Comment:
On current master, all algorithms are using rest api and also rest conversions, even in internal backend. But the upload plan was always the same. If we use the information about the partition location, we get significant speedup in internal backend on rest api ( we transfer the data on the same node)